File: ts/project/.sdk/tm/scala/feature/CacheFeature.scala
package SCALAPACKAGE.feature

import java.util.{ArrayList, LinkedHashMap, List => JList, Map => JMap}
import java.util.function.Supplier
import SCALAPACKAGE.core.{Context, FetcherFn, SdkClient}

// Response caching for safe (read) requests. Wraps the active transport and
// serves a fresh cached snapshot instead of hitting the network when the
// same method+URL was fetched within `ttl` ms (default: 5000). Only
// successful (2xx) responses to cacheable methods (default: GET) are
// stored, keyed by method+URL. The cache is bounded (`max` entries, default
// 256, oldest evicted first) and every hit/miss/bypass is counted. Bodies
// are snapshotted on capture so both the current caller and later hits can
// re-read the JSON body repeatedly.
class CacheFeature extends BaseFeature("cache", "0.0.1", true) {

  private var client: SdkClient = null
  private var options: JMap[String, Object] = null
  private var store: JMap[String, CacheEntry] = null
  private var order: JList[String] = null

  // Activity tracking (mirrors the ts client._cache record).
  var hit: Int = 0
  var miss: Int = 0
  var bypass: Int = 0

  private class CacheEntry {
    var expiry: Long = 0
    var snapshot: CacheSnapshot = null
  }

  private class CacheSnapshot {
    var status: Int = 0
    var statusText: String = ""
    var data: Object = null
    var headers: JMap[String, Object] = new LinkedHashMap[String, Object]()
  }

  override def init(ctx: Context, options: JMap[String, Object]): Unit = {
    this.client = ctx.client
    this.options = options
    this.active = FeatureOptions.foptBool(options, "active", false)

    if (!this.active) {
      return
    }

    this.store = new LinkedHashMap[String, CacheEntry]()
    this.order = new ArrayList[String]()

    val inner: FetcherFn = ctx.utility.fetcher

    ctx.utility.fetcher = (ctx2, url, fetchdef) => through(ctx2, url, fetchdef, inner)
  }

  private def through(ctx: Context, url: String, fetchdef: JMap[String, Object],
      inner: FetcherFn): Object = {

    var method = "GET"
    val m = fetchdef.get("method")
    m match {
      case s: String if s != "" => method = s.toUpperCase()
      case _ =>
    }

    var methods = FeatureOptions.foptStrList(this.options, "methods")
    if (methods == null) {
      methods = java.util.List.of("GET")
    }
    var cacheable = false
    val mit = methods.iterator()
    while (!cacheable && mit.hasNext) {
      val cm = mit.next()
      if (cm.toUpperCase().equals(method)) {
        cacheable = true
      }
    }
    if (!cacheable) {
      return inner(ctx, url, fetchdef)
    }

    val key = method + " " + url
    val now = FeatureOptions.foptNow(this.options).getAsLong()

    val hitEntry = this.store.get(key)
    if (hitEntry != null && hitEntry.expiry > now) {
      this.hit += 1
      return replay(hitEntry.snapshot)
    }

    var res: Object = null
    try {
      res = inner(ctx, url, fetchdef)
    } catch {
      case err: RuntimeException =>
        this.bypass += 1
        throw err
    }

    if (storable(res)) {
      val snapshot = this.snapshot(res)
      val ttl = FeatureOptions.foptInt(this.options, "ttl", 5000)
      evict()
      val entry = new CacheEntry()
      entry.expiry = now + ttl
      entry.snapshot = snapshot
      this.store.put(key, entry)
      this.order.add(key)
      this.miss += 1
      return replay(snapshot)
    }

    this.bypass += 1
    res
  }

  private def storable(res: Object): Boolean = {
    val status = FeatureOptions.fresStatus(res)
    status >= 200 && status < 300
  }

  private def snapshot(res: Object): CacheSnapshot = {
    val rm: JMap[String, Object] = res match {
      case m: JMap[_, _] => m.asInstanceOf[JMap[String, Object]]
      case _ => null
    }

    val snap = new CacheSnapshot()

    val status = FeatureOptions.fresStatus(res)
    if (status >= 0) {
      snap.status = status
    }
    if (rm != null) {
      rm.get("statusText") match {
        case st: String => snap.statusText = st
        case _ =>
      }
      rm.get("json") match {
        case jf: Supplier[_] => snap.data = jf.asInstanceOf[Supplier[Object]].get()
        case _ =>
      }
      rm.get("headers") match {
        case hm0: JMap[_, _] =>
          val hm = hm0.asInstanceOf[JMap[String, Object]]
          val it = hm.entrySet().iterator()
          while (it.hasNext) {
            val h = it.next()
            snap.headers.put(h.getKey.toLowerCase(), h.getValue)
          }
        case _ =>
      }
    }

    snap
  }

  // replay builds a fresh transport-shaped response so the body stays
  // re-readable for every consumer.
  private def replay(snap: CacheSnapshot): JMap[String, Object] = {
    val data = snap.data
    val headers = new LinkedHashMap[String, Object](snap.headers)
    val out = new LinkedHashMap[String, Object]()
    out.put("status", java.lang.Integer.valueOf(snap.status))
    out.put("statusText", snap.statusText)
    out.put("body", "not-used")
    val js: Supplier[Object] = () => data
    out.put("json", js)
    out.put("headers", headers)
    out
  }

  // evict drops oldest entries (FIFO) until the store is under `max`.
  private def evict(): Unit = {
    val max = FeatureOptions.foptInt(this.options, "max", 256)
    while (this.store.size() >= max && !this.order.isEmpty) {
      val oldest = this.order.remove(0)
      this.store.remove(oldest)
    }
  }
}

File: ts/project/.sdk/tm/java/feature/CacheFeature.java
package JAVAPACKAGE.feature;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Supplier;

import JAVAPACKAGE.core.Context;
import JAVAPACKAGE.core.SdkClient;
import JAVAPACKAGE.core.Utility;

// Response caching for safe (read) requests. Wraps the active transport and
// serves a fresh cached snapshot instead of hitting the network when the
// same method+URL was fetched within `ttl` ms (default: 5000). Only
// successful (2xx) responses to cacheable methods (default: GET) are
// stored, keyed by method+URL. The cache is bounded (`max` entries, default
// 256, oldest evicted first) and every hit/miss/bypass is counted. Bodies
// are snapshotted on capture so both the current caller and later hits can
// re-read the JSON body repeatedly.
@SuppressWarnings({"unchecked"})
public class CacheFeature extends BaseFeature {

  private SdkClient client;
  private Map<String, Object> options;
  private Map<String, CacheEntry> store;
  private List<String> order;

  // Activity tracking (mirrors the ts client._cache record).
  public int hit = 0;
  public int miss = 0;
  public int bypass = 0;

  private static final class CacheEntry {
    long expiry;
    CacheSnapshot snapshot;
  }

  private static final class CacheSnapshot {
    int status;
    String statusText = "";
    Object data;
    Map<String, Object> headers = new LinkedHashMap<>();
  }

  public CacheFeature() {
    super("cache", "0.0.1", true);
  }

  @Override
  public void init(Context ctx, Map<String, Object> options) {
    this.client = ctx.client;
    this.options = options;
    this.active = FeatureOptions.foptBool(options, "active", false);

    if (!this.active) {
      return;
    }

    this.store = new LinkedHashMap<>();
    this.order = new ArrayList<>();

    final Utility.FetcherFn inner = ctx.utility.fetcher;

    ctx.utility.fetcher = (ctx2, url, fetchdef) ->
        through(ctx2, url, fetchdef, inner);
  }

  private Object through(Context ctx, String url, Map<String, Object> fetchdef,
      Utility.FetcherFn inner) {

    String method = "GET";
    Object m = fetchdef.get("method");
    if (m instanceof String && !"".equals(m)) {
      method = ((String) m).toUpperCase();
    }

    List<String> methods = FeatureOptions.foptStrList(this.options, "methods");
    if (methods == null) {
      methods = List.of("GET");
    }
    boolean cacheable = false;
    for (String cm : methods) {
      if (cm.toUpperCase().equals(method)) {
        cacheable = true;
        break;
      }
    }
    if (!cacheable) {
      return inner.fetch(ctx, url, fetchdef);
    }

    String key = method + " " + url;
    long now = FeatureOptions.foptNow(this.options).getAsLong();

    CacheEntry hitEntry = this.store.get(key);
    if (hitEntry != null && hitEntry.expiry > now) {
      this.hit++;
      return replay(hitEntry.snapshot);
    }

    Object res;
    try {
      res = inner.fetch(ctx, url, fetchdef);
    }
    catch (RuntimeException err) {
      this.bypass++;
      throw err;
    }

    if (storable(res)) {
      CacheSnapshot snapshot = snapshot(res);
      int ttl = FeatureOptions.foptInt(this.options, "ttl", 5000);
      evict();
      CacheEntry entry = new CacheEntry();
      entry.expiry = now + ttl;
      entry.snapshot = snapshot;
      this.store.put(key, entry);
      this.order.add(key);
      this.miss++;
      return replay(snapshot);
    }

    this.bypass++;
    return res;
  }

  private boolean storable(Object res) {
    int status = FeatureOptions.fresStatus(res);
    return status >= 200 && status < 300;
  }

  private CacheSnapshot snapshot(Object res) {
    Map<String, Object> rm = res instanceof Map ? (Map<String, Object>) res : null;

    CacheSnapshot snap = new CacheSnapshot();

    int status = FeatureOptions.fresStatus(res);
    if (status >= 0) {
      snap.status = status;
    }
    if (rm != null) {
      Object st = rm.get("statusText");
      if (st instanceof String) {
        snap.statusText = (String) st;
      }
      Object jf = rm.get("json");
      if (jf instanceof Supplier) {
        snap.data = ((Supplier<Object>) jf).get();
      }
      Object headers = rm.get("headers");
      if (headers instanceof Map) {
        for (Map.Entry<String, Object> h : ((Map<String, Object>) headers).entrySet()) {
          snap.headers.put(h.getKey().toLowerCase(), h.getValue());
        }
      }
    }

    return snap;
  }

  // replay builds a fresh transport-shaped response so the body stays
  // re-readable for every consumer.
  private Map<String, Object> replay(CacheSnapshot snap) {
    final Object data = snap.data;
    Map<String, Object> headers = new LinkedHashMap<>(snap.headers);
    Map<String, Object> out = new LinkedHashMap<>();
    out.put("status", snap.status);
    out.put("statusText", snap.statusText);
    out.put("body", "not-used");
    out.put("json", (Supplier<Object>) () -> data);
    out.put("headers", headers);
    return out;
  }

  // evict drops oldest entries (FIFO) until the store is under `max`.
  private void evict() {
    int max = FeatureOptions.foptInt(this.options, "max", 256);
    while (this.store.size() >= max && !this.order.isEmpty()) {
      String oldest = this.order.remove(0);
      this.store.remove(oldest);
    }
  }
}
